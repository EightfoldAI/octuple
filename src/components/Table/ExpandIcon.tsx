import React from 'react';
import { Button, ButtonShape, ButtonVariant } from '../Button';
import { IconName } from '../Icon';
import { mergeClasses } from '../../shared/utilities';

import styles from './Styles/table.module.scss';

interface DefaultExpandIconProps<RecordType> {
  expandable: boolean;
  expanded: boolean;
  onExpand: (record: RecordType, e: React.MouseEvent<HTMLElement>) => void;
  record: RecordType;
  disabled?: boolean;
}

function renderExpandIcon(collapseText: string, expandText: string) {
  return function expandIcon<RecordType>({
    onExpand,
    record,
    expanded,
    expandable,
    disabled,
  }: DefaultExpandIconProps<RecordType>) {
    return (
      <Button
        ariaLabel={expanded ? collapseText : expandText}
        classNames={mergeClasses([
          styles.tableRowExpandIcon,
          { [styles.tableRowExpandIconVisuallyHidden]: !expandable },
          { [styles.tableExpandedRow]: expandable && expanded },
        ])}
        disabled={disabled}
        htmlType="button"
        iconProps={{
          path: expanded ? IconName.mdiChevronUp : IconName.mdiChevronDown,
        }}
        onClick={(e: React.MouseEvent<HTMLButtonElement, MouseEvent>) => {
          if (disabled) {
            return;
          }
          onExpand(record, e!);
          e?.stopPropagation();
        }}
        shape={ButtonShape.Round}
        variant={ButtonVariant.Neutral}
      />
    );
  };
}

export default renderExpandIcon;
