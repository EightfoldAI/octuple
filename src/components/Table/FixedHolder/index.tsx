import React, {
    forwardRef,
    useCallback,
    useEffect,
    useContext,
    useMemo,
    useRef,
} from 'react';
import { mergeClasses } from '../../../shared/utilities';
import { fillRef } from '../../../shared/ref';
import ColGroup from '../ColGroup';
import type { ColumnsType, ColumnType } from '../Table.types';
import { FixedHeaderProps } from './FixedHolder.types';
import TableContext from '../Context/TableContext';

import styles from './Table.module.scss';

function useColumnWidth(colWidths: readonly number[], columCount: number) {
    return useMemo(() => {
        const cloneColumns: number[] = [];
        for (let i = 0; i < columCount; i += 1) {
            const val = colWidths[i];
            if (val !== undefined) {
                cloneColumns[i] = val;
            } else {
                return null;
            }
        }
        return cloneColumns;
    }, [colWidths.join('_'), columCount]);
}

const FixedHolder = forwardRef<HTMLDivElement, FixedHeaderProps<unknown>>(
    (
        {
            classNames,
            noData,
            columns,
            flattenColumns,
            colWidths,
            columCount,
            stickyOffsets,
            direction,
            fixHeader,
            stickyTopOffset,
            stickyBottomOffset,
            stickyClassName,
            onScroll,
            maxContentScroll,
            children,
            ...props
        },
        ref
    ) => {
        const { scrollbarSize, isSticky } = useContext(TableContext);

        const combinationScrollBarSize =
            isSticky && !fixHeader ? 0 : scrollbarSize;

        // Pass wheel to scroll event
        const scrollRef = useRef<HTMLDivElement>(null);

        const setScrollRef = useCallback((element: HTMLElement) => {
            fillRef(ref, element);
            fillRef(scrollRef, element);
        }, []);

        useEffect(() => {
            function onWheel(e: WheelEvent) {
                const { currentTarget, deltaX } =
                    e as unknown as React.WheelEvent<HTMLDivElement>;
                if (deltaX) {
                    onScroll({
                        currentTarget,
                        scrollLeft: currentTarget.scrollLeft + deltaX,
                    });
                    e.preventDefault();
                }
            }
            scrollRef.current?.addEventListener('wheel', onWheel);

            return () => {
                scrollRef.current?.removeEventListener('wheel', onWheel);
            };
        }, []);

        // Check if all flattenColumns has width
        const allFlattenColumnsWithWidth = useMemo(
            () => flattenColumns.every((column) => column.width >= 0),
            [flattenColumns]
        );

        // Add scrollbar column
        const lastColumn = flattenColumns[flattenColumns.length - 1];
        const ScrollBarColumn: ColumnType<unknown> & { scrollbar: true } = {
            fixed: lastColumn ? lastColumn.fixed : null,
            scrollbar: true,
            onHeaderCell: () => ({
                className: styles.tableCellScrollbar,
            }),
        };

        const columnsWithScrollbar = useMemo<ColumnsType<unknown>>(
            () =>
                combinationScrollBarSize
                    ? [...columns, ScrollBarColumn]
                    : columns,
            [combinationScrollBarSize, columns]
        );

        const flattenColumnsWithScrollbar = useMemo(
            () =>
                combinationScrollBarSize
                    ? [...flattenColumns, ScrollBarColumn]
                    : flattenColumns,
            [combinationScrollBarSize, flattenColumns]
        );

        // Calculate the sticky offsets
        const headerStickyOffsets = useMemo(() => {
            const { right, left } = stickyOffsets;
            return {
                ...stickyOffsets,
                left:
                    direction === 'rtl'
                        ? [
                              ...left.map(
                                  (width) => width + combinationScrollBarSize
                              ),
                              0,
                          ]
                        : left,
                right:
                    direction === 'rtl'
                        ? right
                        : [
                              ...right.map(
                                  (width) => width + combinationScrollBarSize
                              ),
                              0,
                          ],
                isSticky,
            };
        }, [combinationScrollBarSize, stickyOffsets, isSticky]);

        const mergedColumnWidth = useColumnWidth(colWidths, columCount);

        return (
            <div
                style={{
                    overflow: 'hidden',
                    ...(isSticky
                        ? { top: stickyTopOffset, bottom: stickyBottomOffset }
                        : {}),
                }}
                ref={setScrollRef}
                className={mergeClasses([
                    classNames,
                    { [stickyClassName]: !!stickyClassName },
                ])}
            >
                <table
                    style={{
                        tableLayout: 'fixed',
                        visibility:
                            noData || mergedColumnWidth ? null : 'hidden',
                    }}
                >
                    {(!noData ||
                        !maxContentScroll ||
                        allFlattenColumnsWithWidth) && (
                        <ColGroup
                            colWidths={
                                mergedColumnWidth
                                    ? [
                                          ...mergedColumnWidth,
                                          combinationScrollBarSize,
                                      ]
                                    : []
                            }
                            columCount={columCount + 1}
                            columns={flattenColumnsWithScrollbar}
                        />
                    )}
                    {children({
                        ...props,
                        stickyOffsets: headerStickyOffsets,
                        columns: columnsWithScrollbar,
                        flattenColumns: flattenColumnsWithScrollbar,
                    })}
                </table>
            </div>
        );
    }
);

FixedHolder.displayName = 'FixedHolder';

export default FixedHolder;
