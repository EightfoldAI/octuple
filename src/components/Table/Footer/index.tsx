import React, { useMemo } from 'react';
import Summary from './Summary';
import SummaryContext from './SummaryContext';
import { FooterProps } from './Footer.types';

import styles from '../table.module.scss';

function Footer<RecordType>({
    children,
    stickyOffsets,
    flattenColumns,
}: FooterProps<RecordType>) {
    const lastColumnIndex = flattenColumns.length - 1;
    const scrollColumn = flattenColumns[lastColumnIndex];

    const summaryContext = useMemo(
        () => ({
            stickyOffsets,
            flattenColumns,
            scrollColumnIndex: scrollColumn?.scrollbar ? lastColumnIndex : null,
        }),
        [scrollColumn, flattenColumns, lastColumnIndex, stickyOffsets]
    );

    return (
        <SummaryContext.Provider value={summaryContext}>
            <tfoot className={styles.tableSummary}>{children}</tfoot>
        </SummaryContext.Provider>
    );
}

export default Footer;

export const FooterComponents = Summary;
