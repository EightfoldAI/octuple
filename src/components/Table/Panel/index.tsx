import React from 'react';
import { TitleProps } from './Panel.types';

function Panel({ classNames, children }: TitleProps) {
    return <div className={classNames}>{children}</div>;
}

export default Panel;
