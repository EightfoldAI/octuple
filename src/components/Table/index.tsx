import Table, { TablePaginationConfig } from './Table';

export { ColumnProps } from './Internal/Column';
export {
    ColumnsType,
    ColumnType,
    ColumnGroupType,
    ExpandableConfig,
    FilterConfirmProps,
    FilterValue,
    SorterResult,
    TableProps,
    TableRowSelection,
    TableSize,
} from './Table.types';
export { TablePaginationConfig };

export default Table;
