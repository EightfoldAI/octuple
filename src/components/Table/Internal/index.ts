import OcTable from './OcTable';
import { FooterComponents as Summary } from './Footer';
import Column from './Column';
import ColumnGroup from './ColumnGroup';
import { INTERNAL_COL_DEFINE } from './constant';

export { Summary, Column, ColumnGroup, INTERNAL_COL_DEFINE };

export default OcTable;
