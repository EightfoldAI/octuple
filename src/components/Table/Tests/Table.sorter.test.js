import React from 'react';
import Enzyme, { render, mount } from 'enzyme';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import { eventKeys } from '../../../shared/utilities';
import Table from '../index';

Enzyme.configure({ adapter: new Adapter() });

jest.mock('react', () => ({
  ...jest.requireActual('react'),
  useLayoutEffect: jest.requireActual('react').useEffect,
}));

describe('Table.sorter', () => {
  beforeAll(() => {
    Object.defineProperty(window, 'matchMedia', {
      writable: true,
      value: jest.fn().mockImplementation((query) => ({
        matches: false,
        media: query,
        onchange: null,
        addListener: jest.fn(), // Deprecated
        removeListener: jest.fn(), // Deprecated
        addEventListener: jest.fn(),
        removeEventListener: jest.fn(),
        dispatchEvent: jest.fn(),
      })),
    });
  });

  const sorterFn = (a, b) => a.name[0].charCodeAt() - b.name[0].charCodeAt();

  const column = {
    title: 'Name',
    dataIndex: 'name',
    key: 'name',
    sorter: sorterFn,
  };

  const data = [
    { key: 0, name: 'Jack' },
    { key: 1, name: 'Lucy' },
    { key: 2, name: 'Tom' },
    { key: 3, name: 'Jerry' },
  ];

  function createTable(tableProps, columnProps = {}) {
    return (
      <Table
        columns={[
          {
            ...column,
            ...columnProps,
          },
        ]}
        dataSource={data}
        pagination={false}
        {...tableProps}
      />
    );
  }

  function renderedNames(wrapper) {
    return wrapper.find('BodyRow').map((row) => row.props().record.name);
  }

  it('renders sorter icon correctly', () => {
    const wrapper = render(createTable());
    expect(wrapper.find('thead')).toMatchSnapshot();
  });

  it('default sort order ascend', () => {
    const wrapper = mount(
      createTable(
        {},
        {
          defaultSortOrder: 'ascend',
        }
      )
    );

    expect(renderedNames(wrapper)).toEqual(['Jack', 'Jerry', 'Lucy', 'Tom']);
  });

  it('default sort order descend', () => {
    const wrapper = mount(
      createTable(
        {},
        {
          defaultSortOrder: 'descend',
        }
      )
    );

    const getNameColumn = () => wrapper.find('th').at(0);

    expect(renderedNames(wrapper)).toEqual(['Tom', 'Lucy', 'Jack', 'Jerry']);
    expect(getNameColumn().find('.table-column-sorter-down').exists()).toBeTruthy();
  });

  it('should change aria-sort when default sort order is set to descend', () => {
    const wrapper = mount(
      createTable(
        {
          sortDirections: ['descend', 'ascend'],
        },
        {
          defaultSortOrder: 'descend',
        }
      )
    );

    const getNameColumn = () => wrapper.find('th').at(0);

    // Test that it cycles through the order of sortDirections
    expect(renderedNames(wrapper)).toEqual(['Tom', 'Lucy', 'Jack', 'Jerry']);
    expect(getNameColumn().find('.table-column-sorter-down').exists()).toBeTruthy();

    wrapper.find('.table-column-sorters').simulate('click');
    expect(getNameColumn().find('.table-column-sorter-up').exists()).toBeTruthy();

    wrapper.find('.table-column-sorters').simulate('click');
    expect(getNameColumn().find('.table-column-sorter-up').exists()).toBeFalsy();
    expect(getNameColumn().find('.table-column-sorter-down').exists()).toBeFalsy();
  });

  it('sort records', () => {
    const wrapper = mount(createTable());

    const getNameColumn = () => wrapper.find('th').at(0);
    const getSortButton = () => wrapper.find('.table-column-sorters').at(0);

    // first assert default state
    expect(renderedNames(wrapper)).toEqual(['Jack', 'Lucy', 'Tom', 'Jerry']);
    expect(getNameColumn().find('.table-column-sorter-up').exists()).toBeFalsy();
    expect(getNameColumn().find('.table-column-sorter-down').exists()).toBeFalsy();

    // ascend
    getSortButton().simulate('click');
    expect(renderedNames(wrapper)).toEqual(['Jack', 'Jerry', 'Lucy', 'Tom']);
    expect(getNameColumn().find('.table-column-sorter-up').exists()).toBeTruthy();

    // descend
    getSortButton().simulate('click');
    expect(renderedNames(wrapper)).toEqual(['Tom', 'Lucy', 'Jack', 'Jerry']);
    expect(getNameColumn().find('.table-column-sorter-down').exists()).toBeTruthy();
  });

  it('sort records with keydown', () => {
    const wrapper = mount(createTable());

    // ascend
    wrapper
      .find('.table-column-sorters')
      .at(0)
      .simulate('keydown', { key: eventKeys.ENTER });
    expect(renderedNames(wrapper)).toEqual(['Jack', 'Jerry', 'Lucy', 'Tom']);

    // descend
    wrapper
      .find('.table-column-sorters')
      .at(0)
      .simulate('keydown', { key: eventKeys.ENTER });
    expect(renderedNames(wrapper)).toEqual(['Tom', 'Lucy', 'Jack', 'Jerry']);
  });

  describe('can be controlled by sortOrder', () => {
    it('single', () => {
      const wrapper = mount(
        createTable({
          columns: [{ ...column, sortOrder: 'ascend' }],
        })
      );
      expect(renderedNames(wrapper)).toEqual(['Jack', 'Jerry', 'Lucy', 'Tom']);
    });

    it('invalidate mix with single & multiple sorters', () => {
      const wrapper = mount(
        createTable({
          columns: [
            {
              title: 'Name',
              dataIndex: 'name',
              sortOrder: 'ascend',
              sorter: {
                multiple: 1,
              },
            },
            {
              title: 'Name',
              dataIndex: 'name',
              sortOrder: 'ascend',
              sorter: {},
            },
          ],
        })
      );

      expect(renderedNames(wrapper)).toEqual(['Jack', 'Lucy', 'Tom', 'Jerry']);
    });
  });

  it('provides sortOrder in the sorterFn', () => {
    let actualSortOrder;
    mount(
      createTable(
        {},
        {
          sortOrder: 'ascend',
          sorter: (a, b, sortOrder) => {
            actualSortOrder = sortOrder;
            return sorterFn(a, b);
          },
        }
      )
    );
    expect(actualSortOrder).toEqual('ascend');
  });

  it('can update column sortOrder', () => {
    const wrapper = mount(
      createTable({
        columns: [column],
      })
    );
    expect(renderedNames(wrapper)).toEqual(['Jack', 'Lucy', 'Tom', 'Jerry']);
    wrapper.setProps({
      columns: [{ ...column, sortOrder: 'ascend' }],
    });
    wrapper.update();
    expect(renderedNames(wrapper)).toEqual(['Jack', 'Jerry', 'Lucy', 'Tom']);
  });

  it('fires change event', () => {
    const handleChange = jest.fn();
    const wrapper = mount(createTable({ onChange: handleChange }));

    // ascent
    wrapper.find('.table-column-sorters').simulate('click');
    const sorter1 = handleChange.mock.calls[0][2];
    expect(sorter1.column.dataIndex).toBe('name');
    expect(sorter1.order).toBe('ascend');
    expect(sorter1.field).toBe('name');
    expect(sorter1.columnKey).toBe('name');

    wrapper.find('.table-column-sorters').simulate('click');
    const sorter2 = handleChange.mock.calls[1][2];
    expect(sorter2.column.dataIndex).toBe('name');
    expect(sorter2.order).toBe('descend');
    expect(sorter2.field).toBe('name');
    expect(sorter2.columnKey).toBe('name');

    wrapper.find('.table-column-sorters').simulate('click');
    const sorter3 = handleChange.mock.calls[2][2];
    expect(sorter3.column).toBe(undefined);
    expect(sorter3.order).toBe(undefined);
    expect(sorter3.field).toBe('name');
    expect(sorter3.columnKey).toBe('name');
  });

  it('works with grouping columns in controlled mode', () => {
    const columns = [
      {
        title: 'group',
        key: 'group',
        children: [
          {
            title: 'Name',
            dataIndex: 'name',
            key: 'name',
            sorter: sorterFn,
            sortOrder: 'descend',
          },
          {
            title: 'Age',
            dataIndex: 'age',
            key: 'age',
          },
        ],
      },
    ];
    const testData = [
      { key: 0, name: 'Jack', age: 11 },
      { key: 1, name: 'Lucy', age: 20 },
      { key: 2, name: 'Tom', age: 21 },
      { key: 3, name: 'Jerry', age: 22 },
    ];
    const wrapper = mount(<Table columns={columns} dataSource={testData} />);

    expect(renderedNames(wrapper)).toEqual(['Tom', 'Lucy', 'Jack', 'Jerry']);
  });

  it('Allow column title as render props with sortOrder argument', () => {
    const title = ({ sortOrder }) => (
      <div className="custom-title">{sortOrder}</div>
    );
    const columns = [
      {
        title,
        key: 'group',
        sorter: true,
      },
    ];
    const testData = [
      { key: 0, name: 'Jack', age: 11 },
      { key: 1, name: 'Lucy', age: 20 },
      { key: 2, name: 'Tom', age: 21 },
      { key: 3, name: 'Jerry', age: 22 },
    ];
    const wrapper = mount(<Table columns={columns} dataSource={testData} />);
    expect(wrapper.find('.custom-title').text()).toEqual('');
    wrapper.find('.table-column-sorters').simulate('click');
    expect(wrapper.find('.custom-title').text()).toEqual('ascend');
    wrapper.find('.table-column-sorters').simulate('click');
    expect(wrapper.find('.custom-title').text()).toEqual('descend');
  });

  it('should sort from beginning state when toggle from different columns', () => {
    const columns = [
      {
        title: 'name',
        dataIndex: 'name',
        sorter: true,
      },
      {
        title: 'age',
        dataIndex: 'age',
        sorter: true,
      },
    ];
    const testData = [
      { key: 0, name: 'Jack', age: 11 },
      { key: 1, name: 'Lucy', age: 20 },
      { key: 2, name: 'Tom', age: 21 },
      { key: 3, name: 'Jerry', age: 22 },
    ];
    const wrapper = mount(<Table columns={columns} dataSource={testData} />);

    const getNameColumn = () => wrapper.find('th').at(0);
    const getAgeColumn = () => wrapper.find('th').at(1);
    const getNameSortButton = () => wrapper.find('.table-column-sorters').at(0);
    const getAgeSortButton = () => wrapper.find('.table-column-sorters').at(1);

    // sort name
    getNameSortButton().simulate('click');
    expect(getNameColumn().find('.table-column-sorter-up').exists()).toBeTruthy();
    expect(getAgeColumn().find('.table-column-sorter-up').exists()).toBeFalsy();
    expect(getAgeColumn().find('.table-column-sorter-down').exists()).toBeFalsy();

    // sort age
    getAgeSortButton().simulate('click');
    expect(getNameColumn().find('.table-column-sorter-up').exists()).toBeFalsy();
    expect(getNameColumn().find('.table-column-sorter-down').exists()).toBeFalsy();
    expect(getAgeColumn().find('.table-column-sorter-up').exists()).toBeTruthy();
  });

  it('should toggle sort state when columns are put in render', () => {
    const testData = [
      { key: 0, name: 'Jack', age: 11 },
      { key: 1, name: 'Lucy', age: 20 },
      { key: 2, name: 'Tom', age: 21 },
      { key: 3, name: 'Jerry', age: 22 },
    ];
    class TableTest extends React.Component {
      state = {
        pagination: {},
      };

      onChange = (pagination) => {
        this.setState({ pagination });
      };

      render() {
        const columns = [
          {
            title: 'name',
            dataIndex: 'name',
            sorter: true,
          },
        ];
        const { pagination } = this.state;
        return (
          <Table
            columns={columns}
            pagination={pagination}
            dataSource={testData}
            onChange={this.onChange}
          />
        );
      }
    }

    const wrapper = mount(<TableTest />);

    const getNameColumn = () => wrapper.find('th').at(0);
    const getSortButton = () => wrapper.find('.table-column-sorters').at(0);

    // sort name
    getSortButton().simulate('click');
    expect(getNameColumn().find('.table-column-sorter-up').exists()).toBeTruthy();

    // sort name
    getSortButton().simulate('click');
    expect(getNameColumn().find('.table-column-sorter-down').exists()).toBeTruthy();

    // sort name
    getSortButton().simulate('click');
    expect(getNameColumn().find('.table-column-sorter-up').exists()).toBeFalsy();
    expect(getNameColumn().find('.table-column-sorter-down').exists()).toBeFalsy();
  });

  it('should toggle sort state when columns with non primitive properties are put in render', () => {
    const testData = [
      { key: 0, name: 'Jack', age: 11 },
      { key: 1, name: 'Lucy', age: 20 },
      { key: 2, name: 'Tom', age: 21 },
      { key: 3, name: 'Jerry', age: 22 },
    ];
    class TableTest extends React.Component {
      state = {
        pagination: {},
      };

      onChange = (pagination) => {
        this.setState({ pagination });
      };

      render() {
        const columns = [
          {
            title: 'name',
            dataIndex: 'name',
            sorter: true,
            render: (text) => text,
            array: ['1', '2', 3],
          },
        ];
        const { pagination } = this.state;
        return (
          <Table
            columns={columns}
            pagination={pagination}
            dataSource={testData}
            onChange={this.onChange}
          />
        );
      }
    }

    const wrapper = mount(<TableTest />);

    const getNameColumn = () => wrapper.find('th').at(0);
    const getSortButton = () => wrapper.find('.table-column-sorters').at(0);

    // sort name
    getSortButton().simulate('click');
    expect(getNameColumn().find('.table-column-sorter-up').exists()).toBeTruthy();

    // sort name
    getSortButton().simulate('click');
    expect(getNameColumn().find('.table-column-sorter-down').exists()).toBeTruthy();

    // sort name
    getSortButton().simulate('click');
    expect(getNameColumn().find('.table-column-sorter-up').exists()).toBeFalsy();
    expect(getNameColumn().find('.table-column-sorter-down').exists()).toBeFalsy();
  });

  it('should toggle sort state when columns with key are put in render', () => {
    const testData = [
      { key: 0, name: 'Jack', age: 11 },
      { key: 1, name: 'Lucy', age: 20 },
      { key: 2, name: 'Tom', age: 21 },
      { key: 3, name: 'Jerry', age: 22 },
    ];
    class TableTest extends React.Component {
      state = {
        pagination: {},
      };

      onChange = (pagination) => {
        this.setState({ pagination });
      };

      render() {
        const columns = [
          {
            title: 'name',
            dataIndex: 'name',
            sorter: true,
            key: 'a',
            style: {
              fontSize: 18,
            },
          },
        ];
        const { pagination } = this.state;
        return (
          <Table
            columns={columns}
            pagination={pagination}
            dataSource={testData}
            onChange={this.onChange}
          />
        );
      }
    }

    const wrapper = mount(<TableTest />);
    const getNameColumn = () => wrapper.find('th').at(0);
    const getSortButton = () => wrapper.find('.table-column-sorters').at(0);

    // Initial state
    expect(getNameColumn().find('.table-column-sorter-up').exists()).toBeFalsy();
    expect(getNameColumn().find('.table-column-sorter-down').exists()).toBeFalsy();

    // sort name
    getSortButton().simulate('click');
    expect(getNameColumn().find('.table-column-sorter-up').exists()).toBeTruthy();

    // sort name
    getSortButton().simulate('click');
    expect(getNameColumn().find('.table-column-sorter-down').exists()).toBeTruthy();

    // sort name
    getSortButton().simulate('click');
    expect(getNameColumn().find('.table-column-sorter-up').exists()).toBeFalsy();
    expect(getNameColumn().find('.table-column-sorter-down').exists()).toBeFalsy();
  });

  it('should first sort by descend, then ascend, then cancel sort', () => {
    const wrapper = mount(
      createTable({
        sortDirections: ['descend', 'ascend'],
      })
    );
    const getNameColumn = () => wrapper.find('th').at(0);
    const getSortButton = () => wrapper.find('.table-column-sorters').at(0);

    // descend
    getSortButton().simulate('click');
    expect(renderedNames(wrapper)).toEqual(['Tom', 'Lucy', 'Jack', 'Jerry']);
    expect(getNameColumn().find('.table-column-sorter-down').exists()).toBeTruthy();

    // ascend
    getSortButton().simulate('click');
    expect(renderedNames(wrapper)).toEqual(['Jack', 'Jerry', 'Lucy', 'Tom']);
    expect(getNameColumn().find('.table-column-sorter-up').exists()).toBeTruthy();

    // cancel sort
    getSortButton().simulate('click');
    expect(renderedNames(wrapper)).toEqual(['Jack', 'Lucy', 'Tom', 'Jerry']);
    expect(getNameColumn().find('.table-column-sorter-up').exists()).toBeFalsy();
    expect(getNameColumn().find('.table-column-sorter-down').exists()).toBeFalsy();
  });

  it('should first sort by descend, then cancel sort', () => {
    const wrapper = mount(
      createTable({
        sortDirections: ['descend'],
      })
    );

    const getNameColumn = () => wrapper.find('th').at(0);
    const getSortButton = () => wrapper.find('.table-column-sorters').at(0);

    // default
    expect(getNameColumn().find('.table-column-sorter-up').exists()).toBeFalsy();
    expect(getNameColumn().find('.table-column-sorter-down').exists()).toBeFalsy();

    // descend
    getSortButton().simulate('click');
    expect(renderedNames(wrapper)).toEqual(['Tom', 'Lucy', 'Jack', 'Jerry']);
    expect(getNameColumn().find('.table-column-sorter-down').exists()).toBeTruthy();

    // cancel sort
    getSortButton().simulate('click');
    expect(renderedNames(wrapper)).toEqual(['Jack', 'Lucy', 'Tom', 'Jerry']);
    expect(getNameColumn().find('.table-column-sorter-up').exists()).toBeFalsy();
    expect(getNameColumn().find('.table-column-sorter-down').exists()).toBeFalsy();
  });

  it('should first sort by descend, then cancel sort. (column prop)', () => {
    const wrapper = mount(
      createTable(
        {},
        {
          sortDirections: ['descend'],
        }
      )
    );

    const getNameColumn = () => wrapper.find('th').at(0);
    const getSortButton = () => wrapper.find('.table-column-sorters').at(0);

    // default
    expect(getNameColumn().find('.table-column-sorter-up').exists()).toBeFalsy();
    expect(getNameColumn().find('.table-column-sorter-down').exists()).toBeFalsy();

    // descend
    getSortButton().simulate('click');
    expect(renderedNames(wrapper)).toEqual(['Tom', 'Lucy', 'Jack', 'Jerry']);
    expect(getNameColumn().find('.table-column-sorter-down').exists()).toBeTruthy();

    // cancel sort
    getSortButton().simulate('click');
    expect(renderedNames(wrapper)).toEqual(['Jack', 'Lucy', 'Tom', 'Jerry']);
    expect(getNameColumn().find('.table-column-sorter-up').exists()).toBeFalsy();
    expect(getNameColumn().find('.table-column-sorter-down').exists()).toBeFalsy();
  });

  it('should support onHeaderCell in sort column', () => {
    const onClick = jest.fn();
    const wrapper = mount(
      <Table
        columns={[
          {
            title: 'title',
            onHeaderCell: () => ({ onClick }),
            sorter: true,
          },
        ]}
      />
    );
    wrapper.find('.table-column-sorters').simulate('click');
    expect(onClick).toHaveBeenCalled();
  });

  it('should not being inifinite loop when using Table.Column with sortOrder', () => {
    class Demo extends React.Component {
      componentDidMount() {
        this.setState({});
      }

      render() {
        return (
          <Table dataSource={[]}>
            <Table.Column
              title="Age"
              dataIndex="age"
              sorter
              sortOrder="ascend"
              key="age"
            />
          </Table>
        );
      }
    }
    expect(() => {
      mount(<Demo />);
    }).not.toThrow();
  });

  it('should support defaultOrder in Column', () => {
    const wrapper = mount(
      <Table dataSource={[{ key: '1', age: 1 }]}>
        <Table.Column
          title="Age"
          dataIndex="age"
          sorter
          defaultSortOrder="ascend"
          key="age"
        />
      </Table>
    );
    expect(wrapper.render()).toMatchSnapshot();
  });

  it('invalidate sorter should not display sorter button', () => {
    const wrapper = mount(
      <Table
        columns={[
          {
            title: 'Name',
            dataIndex: 'name',
            key: 'name',
            sorter: false,
          },
          {
            title: 'Age',
            dataIndex: 'age',
            key: 'age',
            sorter: null,
          },
          {
            title: 'Address',
            dataIndex: 'address',
            key: 'address',
            sorter: undefined,
          },
        ]}
      />
    );

    expect(wrapper.find('.table-column-sorter-inner')).toHaveLength(0);
  });

  it('table with sugar column', () => {
    const wrapper = mount(
      <Table>
        <Table.Column
          title="Chinese Score"
          dataIndex="chinese"
          sorter={{
            compare: (a, b) => a.chinese - b.chinese,
            multiple: 3,
          }}
        />
        <Table.Column
          title="Math Score"
          dataIndex="math"
          sorter={{
            compare: (a, b) => a.math - b.math,
            multiple: 2,
          }}
        />
      </Table>
    );

    wrapper.find('.table-column-sorters').first().simulate('click');

    expect(wrapper.find('.table-column-sort')).toHaveLength(1);
  });

  it('surger should support sorterOrder', () => {
    const wrapper = mount(
      <Table>
        <Table.Column
          key="name"
          title="Name"
          dataIndex="name"
          sortOrder="ascend"
          sorter
        />
      </Table>
    );

    expect(
      wrapper
        .find('.table-column-sorter')
        .last()
        .hasClass('table-column-sorter-up')
    ).toBeTruthy();
    expect(
      wrapper
        .find('.table-column-sorter')
        .last()
        .hasClass('table-column-sorter-down')
    ).toBeFalsy();
  });

  it('onChange with correct sorter for multiple', () => {
    const groupColumns = [
      {
        title: 'Math Score',
        dataIndex: 'math',
        sorter: { multiple: 1 },
      },
      {
        title: 'English Score',
        dataIndex: 'english',
        sorter: { multiple: 2 },
      },
    ];

    const groupData = [
      {
        key: '1',
        name: 'John Brown',
        chinese: 98,
        math: 60,
        english: 70,
      },
    ];

    const onChange = jest.fn();
    const wrapper = mount(
      <Table columns={groupColumns} data={groupData} onChange={onChange} />
    );

    function clickToMatchExpect(index, sorter) {
      wrapper.find('.table-column-sorters').at(index).simulate('click');

      expect(onChange).toHaveBeenCalledWith(
        expect.anything(),
        expect.anything(),
        expect.objectContaining(sorter),
        expect.anything()
      );

      onChange.mockReset();
    }

    // First
    clickToMatchExpect(0, { field: 'math', order: 'ascend' });
    clickToMatchExpect(0, { field: 'math', order: 'descend' });
    clickToMatchExpect(0, { field: 'math', order: undefined });

    // Last
    clickToMatchExpect(1, { field: 'english', order: 'ascend' });
    clickToMatchExpect(1, { field: 'english', order: 'descend' });
    clickToMatchExpect(1, { field: 'english', order: undefined });
  });
});
