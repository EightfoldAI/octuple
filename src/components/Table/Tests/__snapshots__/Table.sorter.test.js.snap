version https://git-lfs.github.com/spec/v1
oid sha256:56805733e98c6f9dd81be900dc172f8c17429992bcd2248365a16b339141ecaf
size 843387
